### ISSUE TITLE
panic: send on closed channel

### ISSUE THREAD

**Issue by greysond** [drone/drone#2480]
Getting this panic in the server when a build outputs large logs:

```
panic: send on closed channel

goroutine 168623 [running]:
github.com/drone/drone/server.LogStreamSSE.func2.1(0xc428365a38, 0x1, 0x1)
       /go/src/github.com/drone/drone/server/stream.go:196 +0xc4
created by github.com/drone/drone/vendor/github.com/cncd/logging.(*log).Write
       /go/src/github.com/drone/drone/vendor/github.com/cncd/logging/log.go:74 +0x1cd
```

Possibly related error on the agent just before the panic:
```
grpc error: upload(): code: ResourceExhausted: rpc error: code = ResourceExhausted desc = grpc: trying to send message larger than max (6135644 vs. 4194304)
```

**Comment by jbrockopp** [comment]
We are experiencing the same behavior. I've verified the logs on both on the server and the agent.

IMO this is a big problem because it causes the server to crash which in turn causes an outage.

**Comment by patrickjahns** [comment]
@greysond

What version are you on?

**Comment by greysond** [comment]
v0.8.6

**Comment by jbrockopp** [comment]
Same... We're on `0.8.6`.

**Comment by greysond** [comment]
Sorry, that's pretty relevant!

**Comment by bradrydzewski** [comment]
I pushed a fix to recover on panic. As soon as we verify the change we will tag an 0.8.7 release (hopefully tomorrow).

**Comment by nsheridan** [comment]
I'm running Drone 0.8.8 and I'm seeing panics similar to the original error. These are also in stream.go but in a different location in the file:

```
-sh-4.2$ sudo docker logs --since 24h drone-server 2>&1 |grep -A 5 panic:
panic: send on closed channel

goroutine 1915572 [running]:
github.com/drone/drone/server.EventStreamSSE.func2.1(0x0, 0x0, 0xc421511400, 0x93e, 0xb5f, 0xc422abf050)
	/go/src/github.com/drone/drone/server/stream.go:89 +0x176
created by github.com/drone/drone/vendor/github.com/cncd/pubsub.(*topic).publish
--
panic: send on closed channel

goroutine 292436 [running]:
github.com/drone/drone/server.EventStreamSSE.func2.1(0x0, 0x0, 0xc422188800, 0xae5, 0xb39, 0xc42149d890)
	/go/src/github.com/drone/drone/server/stream.go:89 +0x176
created by github.com/drone/drone/vendor/github.com/cncd/pubsub.(*topic).publish
--
panic: send on closed channel

goroutine 293348 [running]:
github.com/drone/drone/server.EventStreamSSE.func2.1(0x0, 0x0, 0xc423918800, 0xa7c, 0xb9e, 0xc4214b1cb0)
	/go/src/github.com/drone/drone/server/stream.go:89 +0x176
created by github.com/drone/drone/vendor/github.com/cncd/pubsub.(*topic).publish
```

https://github.com/drone/drone/blob/master/server/stream.go#L79-L94

I've created [a PR](https://github.com/drone/drone/pull/2526) which runs a `recover()` similarly to the earlier fix.

**Comment by bradrydzewski** [comment]
closing this as resolved in 1.0
